### PR TITLE
Fix/dashboard gateway

### PR DESCRIPTION
**Pull request by Saif-Alhaider**
separate Remote gateway into multiple gateways
- location gateway
- User gateway
- Restaurant gateway
- Taxi gateway
- removed RemoteGateway class
also create an interface for each gateway
i didn't separate the fake gateway for now

### CODE REVIEW

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.thechance.common.data.remote.gateway.taxi_gateway
```
Comment (mujtaba912) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt:1]: the gateway interface must be in domain layer and its implementation must be in data layer you broke the architecture

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.thechance.common.domain.entity.AddTaxi
+import org.thechance.common.domain.entity.DataWrapper
+import org.thechance.common.domain.entity.Taxi
+
+interface ITaxiGateway {
+    suspend fun getTaxis(): DataWrapper<Taxi>
+    suspend fun createTaxi(taxi: AddTaxi): Taxi
```
Comment (mujtaba912) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt:9]: AddTaxi  data class name changed into NewTaxiInfo so taxi should be of NewTaxiInfo type and not AddTaxi

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/TaxiGateway.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.thechance.common.domain.entity.AddTaxi
+import org.thechance.common.domain.entity.DataWrapper
+import org.thechance.common.domain.entity.Taxi
+
+class TaxiGateway:ITaxiGateway {
+    override suspend fun getTaxis(): DataWrapper<Taxi> {
+        TODO("Not yet implemented")
+    }
+
+    override suspend fun createTaxi(taxi: AddTaxi): Taxi {
```
Comment (mujtaba912) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/TaxiGateway.kt:12]: AddTaxi  data class name changed into NewTaxiInfo so taxi should be of NewTaxiInfo type and not AddTaxi

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/user_gateway/IUsersGateway.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.thechance.common.data.remote.gateway.user_gateway
```
Comment (mujtaba912) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/user_gateway/IUsersGateway.kt:1]: the gateway interface must be in domain layer and its implementation must be in data layer you broke the architecture

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/location_gateway/ILocationGateway.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.thechance.common.data.remote.gateway.location_gateway
```
[the remark this answers is not in the document]
Comment (Asiasama710) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/location_gateway/ILocationGateway.kt:1]: I agree

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt (hunk 2 above)
Comment (Asiasama710) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt:9]: just change addTaxi to NewTaxiInfo that  already exist

Comment (Asiasama710) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt:9]: image

**Review by Asiasama710 — CHANGES_REQUESTED** [review]

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/TryToExecute.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import io.ktor.client.plugins.ClientRequestException
+import io.ktor.client.statement.HttpResponse
+import org.thechance.common.data.remote.model.ServerResponse
+import org.thechance.common.domain.entity.InvalidCredentialsException
+import org.thechance.common.domain.entity.NoInternetException
+import org.thechance.common.domain.entity.UnknownErrorException
+import org.thechance.common.domain.entity.UserNotFoundException
+import java.net.ConnectException
+
+suspend inline fun <reified T> tryToExecute(
```
Comment (Asiasama710) [inline client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/TryToExecute.kt]: i think u can add it in BaseGateway?

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/location_gateway/ILocationGateway.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Saif-Alhaider) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/location_gateway/ILocationGateway.kt]: resolved 👍

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt (hunk 2 above)
Comment (Saif-Alhaider) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/ITaxiGateway.kt]: resolved 👍

File: client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/TaxiGateway.kt (hunk 3 above)
Comment (Saif-Alhaider) [reply client_dashboard/common/src/commonMain/kotlin/org/thechance/common/data/remote/gateway/taxi_gateway/TaxiGateway.kt]: resolved 👍

**Review by Asiasama710 — APPROVED** [review]

**Review by mujtaba912 — APPROVED** [review]

**Pull request merged**